[Serialization] Errors when referencing subtype named "Type"

NachoSoto opened the pull request:
### Update:
Thanks to @lattner for fixing this in https://github.com/apple/swift/commit/e9d1857ca357defa4836b97d2a76270528825f0a. I rebased and updated the tests to reflect the new diagnostics.

### Old description
*First Swift PR, so don't hesitate to yell if I did something wrong*

I haven't been able to fix this myself, so this *WIP* PR only includes a few failing tests.

```bash
$ swift/utils/build-script -R && llvm/utils/lit/lit.py -sv build/Ninja-ReleaseAssert/swift-macosx-x86_64/test-macosx-x86_64/decl

.../swift/test/decl/ext/extensions.swift:53:1: error: unexpected error produced: cannot extend a metatype 'S1_2.Type'
extension S1_2.Type {} // no-error

.../swift/test/decl/nested.swift:93:18: error: unexpected error produced: type of expression is ambiguous without more context
let s: S.Type = .A
```

Looks like the first error is generated in `createProtocolGenericParams` in `TypeChecker.cpp`, so I looked at how `ModuleFile::getType` is implemented and I'm not sure why it would detect the type as `METATYPE_TYPE` instead of `PAREN_TYPE` (given that that `case` is first), so I imagine the root of the issue might be somewhere else.

It's possible that this problem exists with other types of declarations, I'll add more tests if I can think of other examples.

I'm kind of stuck so I'd appreciate any pointers! I'm also not entirely sure whether this can be fixed 100%, given that some declarations would always be ambiguous, but I think at least the first example should work, and the second one should emit a better diagnostic. Alternatively maybe `Swift` should just emit an error when defining the `Type` subtype?

cwillmor commented:
The type `S1_2.Type` is parsed as a `MetatypeTypeRepr` regardless of whether `S1_2` has a member named `Type` (see `Parser::parseTypeSimple()` in `lib/Parse/ParseType.cpp`). I'm inclined to say that we shouldn't allow `Type` and `Protocol` as non-internal decl names in type context, precisely because they can never be referred to via the `T.Type` syntax.

NachoSoto commented:
> The type `S1_2.Type` is parsed as a `MetatypeTypeRepr` [...] (see `Parser::parseTypeSimple()` in `lib/Parse/ParseType.cpp`

Thanks!

>  I'm inclined to say that we shouldn't allow `Type` and `Protocol` as non-internal decl names in type context, precisely because they can never be referred to via the `T.Type` syntax.

I agree. I'll try to tackle this and follow up here with any questions I'll have :+1:

NachoSoto commented:
@lattner do you think https://github.com/apple/swift/commit/e9d1857ca357defa4836b97d2a76270528825f0a fixes this too? I can rebase and see if these tests pass now as soon as I get to my computer.

lattner commented:
Ah yes, I think it does!  I'm sorry I didn't see this PR before.  If you're interested, it would be great to merge this test case so the error message is tracked.

NachoSoto commented:
@lattner indeed it does! That's no problem, I never found time myself to attempt to fix this.

I rebased and updated the expected errors. Notice that the `type member may not be named 'Type', since it would conflict with the 'foo.Type' expression` diagnostic seems to be emitted twice though, but tests pass now :)

NachoSoto commented on test/decl/ext/extensions.swift:
```diff
@@ -46,6 +46,20 @@ extension S1 {} // no-error
 extension S1.Type {} // expected-error {{cannot extend a metatype 'S1.Type'}}
 extension S1.NestedStruct {} // no-error
 
+struct S1_2 {
```
Let me know if you'd rather me remove these since it's already tested elsewhere.

lattner commented:
Test case looks great, thank you!  I'll take a look at why the diagnostics are getting emitted multiple times when I get a chance.

NachoSoto commented:
Thank you!
